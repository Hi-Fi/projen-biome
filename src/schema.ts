import { writeFileSync } from 'node:fs'
import { join } from 'node:path'
import { compileFromFile } from 'json-schema-to-typescript'

export const createTypes = async (
  typesFile = join(__dirname, '..', 'src', 'biome-configuration.ts'),
) => {
  const types = (
    await compileFromFile(
      join(
        __dirname,
        '..',
        'node_modules',
        '@biomejs',
        'biome',
        'configuration_schema.json',
      ),
      {
        customName: (_schema, keyName): string | undefined => {
          const name = keyName ?? _schema.title ?? _schema.$id
          if (!name) {
            return undefined
          }
          return `I${name}`
        },
      },
    )
  )
    // JSII doesn't like optional values that can also be null
    .replace(/(\?:.*) \| null;/g, '$1;')
    // Removing or undefined, as that should be optional in TS sense like it is marked also
    .replace(/ \| undefined/g, '')
    // JSII wants interfaces with PascalCase
    .replace(/IA11Y/g, 'IA11y')
    // Change type to be OK for JSII
    .replace('boolean | [number, ...number[]]', 'boolean | number[]')
    .replace('$schema', '//$schema')
  // JSII doesn't like optional values that can also be null
  writeFileSync(typesFile, types)
}
