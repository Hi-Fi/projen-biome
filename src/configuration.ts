import { IConfiguration } from './biome-configuration'

export interface BiomeOptions {
  /**
   * Version of Biome to use
   *
   * @default "^1"
   */
  readonly version?: string
  /**
   * Enable linting. Replaces Eslint.
   *
   * @default true
   */
  readonly linter?: boolean
  /**
   * Enable code formatter. Replaces mainly Prettier
   *
   * @default false
   */
  readonly formatter?: boolean
  /**
   * Enable import sorting/organizing. Replaces mainly Prettier
   *
   * @default false
   */
  readonly organizeImports?: boolean

  /**
   * Should arrays be merged or overwritten when creating Biome configuration
   *
   * By default arrays are merged and duplicate values are removed
   *
   * @default true
   */
  readonly mergeArraysInConfiguration?: boolean
  /**
   * Full Biome configuration. Note that this configuration dictates the final outcome is value is set.
   *
   * @example if linter is disabled on main level, it can be enabled on fullConfiguration.formatter.enabled.
   */
  readonly overrides?: IConfiguration
}

/**
 * Default configuration that's used when Eslint was enabled in Projen.
 *
 * Note that Eslint had some formatting options, but those are removed from here and available only within formatter.
 *
 * Generated using `biome migrate eslint`. Some common settings not included here which are in both (Prettier and Eslint) migration configuration
 */
const eslintConfiguration = (projenrcFile?: string): IConfiguration => {
  const include = projenrcFile ? [projenrcFile] : []

  return {
    linter: {
      enabled: true,
      rules: {
        recommended: false,
        complexity: {
          useLiteralKeys: 'error',
        },
        style: {
          useBlockStatements: 'error',
        },
      },
      ignore: [
        '**/*.js',
        '**/*.d.ts',
        '**/node_modules/',
        '**/*.generated.ts',
        '**/coverage',
      ],
    },
    overrides: [
      {
        include,
        linter: {
          rules: {},
        },
      },
    ],
  }
}

/**
 * Default configuration that's used when Prettier was enabled in Projen.
 *
 * Generated using `biome migrate prettier`. Some common settings not included here which are in both (Prettier and Eslint) migration configuration
 */
const prettierConfiguration = (): IConfiguration => ({
  formatter: {
    enabled: true,
    useEditorconfig: true,
    formatWithErrors: true,
    indentStyle: 'space',
    indentWidth: 2,
    lineEnding: 'lf',
    lineWidth: 80,
    attributePosition: 'auto',
    bracketSpacing: true,
  },
  organizeImports: {
    enabled: true,
  },
  linter: {
    enabled: false, // Separating clearly linter and formatter
  },
  javascript: {
    formatter: {
      jsxQuoteStyle: 'double',
      quoteProperties: 'asNeeded',
      trailingCommas: 'all',
      semicolons: 'asNeeded',
      arrowParentheses: 'always',
      bracketSameLine: false,
      quoteStyle: 'single',
      attributePosition: 'auto',
      bracketSpacing: true,
    },
  },
})

/**
 * Merge 2 objects deeply.
 *
 * Can't type to actual type, as function is called recursively.
 *
 * Note that explicit undefined would override value from target; missing key (implicit undefined) is not affecting to result.
 *
 * @param target Object to start with
 * @param overrides Object that can override values in target
 * @param mergeArrays Should arrays be merged or overwritten from overrides. Defaults to merging.
 */
const mergeConfigurations = (
  target: any,
  overrides: any,
  mergeArrays = true,
): any => {
  const results = structuredClone(target)

  for (const key in overrides) {
    // Check if override's key is it's own or inherited; we want to handle only it's own.
    if ((overrides as Object).hasOwnProperty(key)) {
      // Handle arrays
      if (overrides[key] instanceof Array && results[key] instanceof Array) {
        if (mergeArrays) {
          // Merge arrays and drop duplicates
          results[key] = [...new Set([...overrides[key], ...results[key]])]
        } else {
          results[key] = overrides[key]
        }
      } else if (
        overrides[key] instanceof Object &&
        results[key] instanceof Object
      ) {
        // Handle objects
        results[key] = mergeConfigurations(
          results[key],
          overrides[key],
          mergeArrays,
        )
      } else {
        // Handle primitive values
        results[key] = overrides[key]
      }
    }
  }

  return results
}

const commonConfiguration: IConfiguration = {
  vcs: { enabled: false, clientKind: 'git', useIgnoreFile: false },
  files: {
    ignoreUnknown: false,
    ignore: [],
  },
}

export const createBiomeConfiguration = (
  options: BiomeOptions,
  projenrcFile?: string,
): IConfiguration => {
  const defaultLinter: IConfiguration = options.linter
    ? eslintConfiguration(projenrcFile)
    : {}
  const defaultFormatter: IConfiguration = options.formatter
    ? prettierConfiguration()
    : {}
  const defaultOrganized: IConfiguration = options.organizeImports
    ? { organizeImports: { enabled: true } }
    : { organizeImports: { enabled: false } }

  const defaultConfig = mergeConfigurations(
    mergeConfigurations(
      mergeConfigurations(defaultFormatter, defaultLinter),
      defaultOrganized,
    ),
    commonConfiguration,
  )

  return mergeConfigurations(
    defaultConfig,
    options.overrides,
    options.mergeArraysInConfiguration,
  )
}
