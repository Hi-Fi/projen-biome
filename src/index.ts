import { Component, JsonFile, ProjenrcFile } from 'projen'
import { Eslint, type NodeProject, Prettier } from 'projen/lib/javascript'
import { TypeScriptProject } from 'projen/lib/typescript'
import { IConfiguration } from './biome-configuration'
import { BiomeOptions, createBiomeConfiguration } from './configuration'
export * from './biome-configuration'
export { BiomeOptions } from './configuration'

export class Biome extends Component {
  public static of(project: NodeProject): Biome | undefined {
    const isBiome = (c: Component): c is Biome => c instanceof Biome
    return project.components.find(isBiome)
  }

  private readonly configFile: string
  private readonly optionsWithDefaults: BiomeOptions
  private readonly biomeConfiguration: IConfiguration
  private readonly projenrcFile?: string

  constructor(project: NodeProject, options: BiomeOptions = {}) {
    super(project)
    this.configFile = 'biome.jsonc'
    this.projenrcFile = this.project.components.find(
      (component) => component instanceof ProjenrcFile,
    )?.filePath
    this.optionsWithDefaults = {
      mergeArraysInConfiguration: (options as Object).hasOwnProperty(
        'mergeArraysInConfiguration',
      )
        ? options.mergeArraysInConfiguration
        : true,
      formatter: options.formatter ?? false,
      linter: options.linter ?? true,
      organizeImports: options.organizeImports ?? false,
      version: options.version ?? '^1',
      overrides: options.overrides,
    }

    project.addDevDeps(`@biomejs/biome@${this.optionsWithDefaults.version}`)

    this.biomeConfiguration = createBiomeConfiguration(this.optionsWithDefaults)

    new JsonFile(this, this.configFile, {
      obj: this.biomeConfiguration,
      committed: true,
      allowComments: true,
      marker: true,
    })

    const localTask = this.createLocalBiomeTask()
    project.testTask.spawn(localTask)
  }

  public override preSynthesize(): void {
    super.preSynthesize()

    for (const component of this.project.components) {
      if (
        component instanceof Eslint &&
        this.biomeConfiguration.linter?.enabled
      ) {
        throw new Error('Biome linter should not be used together with Eslint')
      }
      if (
        component instanceof Prettier &&
        (this.biomeConfiguration.formatter?.enabled ||
          this.biomeConfiguration.organizeImports?.enabled)
      ) {
        throw new Error(
          'Biome formatter should not be used together with Prettier',
        )
      }
    }
  }

  private createLocalBiomeTask() {
    const targetDirs: string[] = []
    this.projenrcFile && targetDirs.push(this.projenrcFile)

    if (this.project instanceof TypeScriptProject) {
      targetDirs.push(this.project.srcdir)
      targetDirs.push(this.project.testdir)
    }

    return this.project.addTask('biome', {
      description: 'Runs Biome against the codebase',
      steps: [
        {
          exec: `biome check --write ${targetDirs.join(' ')}`,
        },
      ],
    })
  }
}
