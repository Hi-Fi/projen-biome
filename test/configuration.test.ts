import { describe, expect, it } from 'vitest'
import { createBiomeConfiguration } from '../src/configuration'

describe('Configuration', () => {
  describe('should match snapshot', () => {
    describe('without overrides', () => {
      it('without options', () => {
        expect(createBiomeConfiguration({})).toMatchSnapshot()
      })
      it('with linter', () => {
        expect(createBiomeConfiguration({ linter: true })).toMatchSnapshot()
      })

      it('with formatter', () => {
        expect(createBiomeConfiguration({ formatter: true })).toMatchSnapshot()
      })

      it('with import organizer', () => {
        expect(
          createBiomeConfiguration({ organizeImports: true }),
        ).toMatchSnapshot()
      })

      it('with all features', () => {
        expect(
          createBiomeConfiguration({
            linter: true,
            formatter: true,
            organizeImports: true,
          }),
        ).toMatchSnapshot()
      })
    })
  })

  describe('should', () => {
    describe('merge arrays', () => {
      it('when mergeArraysInConfiguration is true', () => {
        const linterConfig = createBiomeConfiguration({ linter: true })

        expect(linterConfig.linter?.ignore?.[0]).toBeDefined()

        const mergedConfig = createBiomeConfiguration({
          linter: true,
          overrides: {
            linter: {
              ignore: ['**/bin'],
            },
          },
        })

        expect(mergedConfig.linter?.ignore?.length).toBe(
          linterConfig.linter!.ignore!.length + 1,
        )
      })

      it('without duplicate values', () => {
        const linterConfig = createBiomeConfiguration({ linter: true })

        expect(linterConfig.linter?.ignore?.[0]).toBeDefined()
        const mergedConfig = createBiomeConfiguration({
          linter: true,
          overrides: {
            linter: {
              ignore: [linterConfig.linter!.ignore![0]],
            },
          },
        })

        expect(
          mergedConfig.linter?.ignore?.filter(
            (ignore) => ignore === linterConfig.linter?.ignore?.[0],
          ).length,
        ).toBe(1)
      })
    })

    describe('replace whole array', () => {
      it('when mergeArraysInConfiguration is false', () => {
        const linterConfig = createBiomeConfiguration({ linter: true })
        if (!linterConfig.linter?.ignore?.[0]) {
        }

        expect(linterConfig.linter?.ignore?.[0]).toBeDefined()
        const mergedConfig = createBiomeConfiguration({
          linter: true,
          mergeArraysInConfiguration: false,
          overrides: {
            linter: {
              ignore: [linterConfig.linter!.ignore![0]],
            },
          },
        })

        expect(
          mergedConfig.linter?.ignore?.filter(
            (ignore) => ignore === linterConfig.linter?.ignore?.[0],
          ).length,
        ).toBe(1)
      })
    })

    it('make deep merge', () => {
      const mergedConfig = createBiomeConfiguration({
        linter: true,
        overrides: {
          linter: {
            rules: {
              a11y: {
                noBlankTarget: 'error',
              },
            },
          },
        },
      })

      expect(mergedConfig.linter?.rules).toBeDefined()
      expect(Object.keys(mergedConfig.linter!.rules!).length).toBeGreaterThan(1)
    })
  })
})
