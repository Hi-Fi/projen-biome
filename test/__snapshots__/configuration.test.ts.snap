// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Configuration > should match snapshot > without overrides > with all features 1`] = `
{
  "files": {
    "ignore": [],
    "ignoreUnknown": false,
  },
  "formatter": {
    "attributePosition": "auto",
    "bracketSpacing": true,
    "enabled": true,
    "formatWithErrors": true,
    "indentStyle": "space",
    "indentWidth": 2,
    "lineEnding": "lf",
    "lineWidth": 80,
    "useEditorconfig": true,
  },
  "javascript": {
    "formatter": {
      "arrowParentheses": "always",
      "attributePosition": "auto",
      "bracketSameLine": false,
      "bracketSpacing": true,
      "jsxQuoteStyle": "double",
      "quoteProperties": "asNeeded",
      "quoteStyle": "single",
      "semicolons": "asNeeded",
      "trailingCommas": "all",
    },
  },
  "linter": {
    "enabled": true,
    "ignore": [
      "**/*.js",
      "**/*.d.ts",
      "**/node_modules/",
      "**/*.generated.ts",
      "**/coverage",
    ],
    "rules": {
      "complexity": {
        "useLiteralKeys": "error",
      },
      "recommended": false,
      "style": {
        "useBlockStatements": "error",
      },
    },
  },
  "organizeImports": {
    "enabled": true,
  },
  "overrides": [
    {
      "include": [],
      "linter": {
        "rules": {},
      },
    },
  ],
  "vcs": {
    "clientKind": "git",
    "enabled": false,
    "useIgnoreFile": false,
  },
}
`;

exports[`Configuration > should match snapshot > without overrides > with formatter 1`] = `
{
  "files": {
    "ignore": [],
    "ignoreUnknown": false,
  },
  "formatter": {
    "attributePosition": "auto",
    "bracketSpacing": true,
    "enabled": true,
    "formatWithErrors": true,
    "indentStyle": "space",
    "indentWidth": 2,
    "lineEnding": "lf",
    "lineWidth": 80,
    "useEditorconfig": true,
  },
  "javascript": {
    "formatter": {
      "arrowParentheses": "always",
      "attributePosition": "auto",
      "bracketSameLine": false,
      "bracketSpacing": true,
      "jsxQuoteStyle": "double",
      "quoteProperties": "asNeeded",
      "quoteStyle": "single",
      "semicolons": "asNeeded",
      "trailingCommas": "all",
    },
  },
  "linter": {
    "enabled": false,
  },
  "organizeImports": {
    "enabled": false,
  },
  "vcs": {
    "clientKind": "git",
    "enabled": false,
    "useIgnoreFile": false,
  },
}
`;

exports[`Configuration > should match snapshot > without overrides > with import organizer 1`] = `
{
  "files": {
    "ignore": [],
    "ignoreUnknown": false,
  },
  "organizeImports": {
    "enabled": true,
  },
  "vcs": {
    "clientKind": "git",
    "enabled": false,
    "useIgnoreFile": false,
  },
}
`;

exports[`Configuration > should match snapshot > without overrides > with linter 1`] = `
{
  "files": {
    "ignore": [],
    "ignoreUnknown": false,
  },
  "linter": {
    "enabled": true,
    "ignore": [
      "**/*.js",
      "**/*.d.ts",
      "**/node_modules/",
      "**/*.generated.ts",
      "**/coverage",
    ],
    "rules": {
      "complexity": {
        "useLiteralKeys": "error",
      },
      "recommended": false,
      "style": {
        "useBlockStatements": "error",
      },
    },
  },
  "organizeImports": {
    "enabled": false,
  },
  "overrides": [
    {
      "include": [],
      "linter": {
        "rules": {},
      },
    },
  ],
  "vcs": {
    "clientKind": "git",
    "enabled": false,
    "useIgnoreFile": false,
  },
}
`;

exports[`Configuration > should match snapshot > without overrides > without options 1`] = `
{
  "files": {
    "ignore": [],
    "ignoreUnknown": false,
  },
  "organizeImports": {
    "enabled": false,
  },
  "vcs": {
    "clientKind": "git",
    "enabled": false,
    "useIgnoreFile": false,
  },
}
`;
